src/Fixer/Comment/HeaderCommentFixer.php

```diff
@@ -125,62 +148,188 @@ private function findHeaderCommentIndex(Tokens $tokens)
+        if ($lineBreakCount < $expectedLineCount) {
+            // because of the way the insert index was determined for header comment there cannot be an empty token here
+            $tokens->insertAt($headerIndex, new Token(array(T_WHITESPACE, str_repeat("\n", $expectedLineCount - $lineBreakCount))));
         }
+    }
 
-        $headCommentTokens = array(
-            new Token(array(T_WHITESPACE, "\n")),
-            new Token(array(T_COMMENT, $this->headerComment)),
-            new Token(array(T_WHITESPACE, "\n\n")),
-        );
+    /**
+     * @param Tokens $tokens
+     * @param int    $indexStart
+     * @param int    $indexEnd
+     *
+     * @return int
+     */
+    private function getLineBreakCount(Tokens $tokens, $indexStart, $indexEnd)
+    {
+        $lineCount = 0;
+        for ($i = $indexStart; $i < $indexEnd; ++$i) {
+            $lineCount += substr_count($tokens[$i]->getContent(), "\n");
+        }
 
-        $newHeaderIndex = null !== $oldHeaderIndex
-            ? $oldHeaderIndex + 1
-            : $this->findHeaderCommentInsertionIndex($tokens) - 1
-        ;
+        return $lineCount;
+    }
 
-        $tokens->overrideRange(1, $newHeaderIndex, $headCommentTokens);
+    /**
+     * @param Tokens $tokens
+     * @param int    $index
+     */
+    private function insertHeader(Tokens $tokens, $index)
+    {
+        $tokens->insertAt($index, new Token(array(self::HEADER_COMMENT === $this->headerCommentType ? T_COMMENT : T_DOC_COMMENT, $this->headerComment)));
     }
 
     /**
-     * Set the desired header text.
+     * @param array|null $configuration
      *
-     * The given text will be trimmed and enclosed into a multiline comment.
-     * If the text is empty, when a file get fixed, the header comment will be
-     * erased.
-     *
-     * @param string $header
+     * @return array
      */
-    private function setHeader($header)
+    private function parseConfiguration(array $configuration = null)
     {
-        $this->header = trim($header);
-        $this->headerComment = '';
+        if (null === $configuration || !array_key_exists('header', $configuration)) {
+            throw new InvalidFixerConfigurationException($this->getName(), 'Configuration is required.');
+        }
+
+        $header = $configuration['header'];
+        if (!is_string($header)) {
+            throw new InvalidFixerConfigurationException($this->getName(), sprintf('Header configuration is invalid. Expected "string", got "%s".', is_object($header) ? get_class($header) : gettype($header)));
+        }
 
-        if ('' !== $this->header) {
-            $this->headerComment = $this->encloseTextInComment($this->header);
+        if (array_key_exists('commentType', $configuration)) {
+            $commentType = $configuration['commentType'];
+            switch ($commentType) {
+                case 'PHPDoc':
+                    $commentType = self::HEADER_PHPDOC;
```

Comment:
my idea:

```php
const HEADER_PHPDOC = 'PHPDoc';
const HEADER_COMMENT = 'comment';

$commentType = $someUserInput;
if (!in_array($commentType, [self::HEADER_PHPDOC, self::HEADER::COMMENT], true) {
    throw Exception();
}
```

and no need for remapping input to internal notation

----------------------

I see no value in keeping value of constant as bits, it's not ANSI C